Fix GPTQ Marlin MoE mixed precision support

chriswritescode-dev opened the pull request:
Fixes several critical issues preventing GPTQ Marlin MoE from working correctly with 8-bit quantization:

1. Shared memory calculation bug: The marlin_moe_wna16 kernel incorrectly calculated sh_b_size for 8-bit quantization, causing memory corruption. Fixed by accounting for b_thread_vecs based on bit width.

2. Workspace allocation: Added bit-width-aware workspace allocation (max_blocks_per_sm = 8 for 8-bit, 4 for 4-bit).

3. Cache validation: Added proper size validation for intermediate_cache2 and intermediate_cache13 buffers to prevent out-of-bounds access.

4. Quantization metadata: Enhanced GPTQMarlinMoEMethod to properly initialize and track per-layer quantization configuration including pack_factor and bit width.

5. SM 12.0 support: Added Blackwell architecture support to scaled_mm in CMakeLists.txt.

Includes new test coverage for mixed precision GPTQ Marlin MoE.

<!-- markdownlint-disable -->

## Purpose
Resole issue #26558 CUDA illegal memory access in fused_marlin_moe with GLM-4.6-GPTQ MoE model (V1 engine, TP=4) 

## Test Plan

### Test 1: test_gptq_marlin_moe_method_uses_dynamic_bits_override
                                                                                                                               
  Dynamic quantization override (8-bit) for layers 0-2  
                                                                                                                               
     • Validates that quant_metadata["bits"] = 8
     • Validates that pack_factor = 4 (32 // 8)
     • Validates that weight shapes are correct (64 // 4 = 16)
### Test 2: test_gptq_marlin_moe_method_retains_base_bits_without_override
                                                                                                                               
    Base 4-bit quantization for layers without override (layer 5)
                                                                                                                               
     • Validates that quant_metadata["bits"] = 4
     • Validates that pack_factor = 8 (32 // 4)
     • Validates that weight shapes are correct (64 // 8 = 8)  

## Test Result
tests/quantization/test_gptq_marlin_moe_mixed_precision.py::test_gptq_marlin_moe_method_uses_dynamic_bits_override PASSED    [ 50%]
tests/quantization/test_gptq_marlin_moe_mixed_precision.py::test_gptq_marlin_moe_method_retains_base_bits_without_override PASSED [100%]



---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

Commit:
Fix GPTQ Marlin MoE mixed precision support

Fixes several critical issues preventing GPTQ Marlin MoE from working correctly with 8-bit quantization:

1. Shared memory calculation bug: The marlin_moe_wna16 kernel incorrectly calculated sh_b_size for 8-bit quantization, causing memory corruption. Fixed by accounting for b_thread_vecs based on bit width.

2. Workspace allocation: Added bit-width-aware workspace allocation (max_blocks_per_sm = 8 for 8-bit, 4 for 4-bit).

3. Cache validation: Added proper size validation for intermediate_cache2 and intermediate_cache13 buffers to prevent out-of-bounds access.

4. Quantization metadata: Enhanced GPTQMarlinMoEMethod to properly initialize and track per-layer quantization configuration including pack_factor and bit width.

5. SM 12.0 support: Added Blackwell architecture support to scaled_mm in CMakeLists.txt.

Includes new test coverage for mixed precision GPTQ Marlin MoE.

Signed-off-by: CS <chris@chriswritescode.dev>

Commit:
Remove redundant override_config call from get_moe_quant_method

Centralize configuration override logic in _initialize_quant_metadata to avoid redundant processing and improve maintainability.

Signed-off-by: CS <chris@chriswritescode.dev>

Commit:
Fix mypy errors for _layer_prefix attribute

Add type: ignore comments for dynamically set _layer_prefix attribute to satisfy mypy type checking.

Signed-off-by: CS <chris@chriswritescode.dev>

Commit:
Apply pre-commit formatting fixes

- Format CUDA code (clang-format)
- Replace setattr with direct assignment (ruff)

Signed-off-by: CS <chris@chriswritescode.dev>

yewentao256 reviewed (commented):
Let's run ci and see what can we get

joninco commented:
## Test Failure Analysis: `test_async_llm_dp.py[True-ray-RequestOutputKind.DELTA]`

  This PR is being blocked by a **pre-existing bug** in vLLM that was recently exposed. The failure is **not caused by changes in this PR**.

  ---

  ### 🔍 The Problem

  The test `v1/distributed/test_async_llm_dp.py::test_load[True-ray-RequestOutputKind.DELTA]` fails with:

  ValueError: Currently, async scheduling only supports mp or uni
  distributed executor backend, but you choose ray.

  **Root Cause:** Incorrect logic in `vllm/config/parallel.py` lines 533-538 conflates two **independent** configuration options:

  | Config Option | Purpose |
  |--------------|---------|
  | `data_parallel_backend` | How to spawn **DP engine processes** (Ray actors vs multiprocessing) |
  | `distributed_executor_backend` | How to spawn **TP/PP workers** within each DP rank (Ray vs mp vs uni) |

  These are separate concerns and should be independently configurable. However, the current code incorrectly forces them to match:

  ```python
  elif self.data_parallel_backend == "ray":
      logger.info("Using ray distributed inference because data_parallel_backend is ray")
      backend = "ray"  # ❌ BUG: Forces distributed_executor_backend="ray"

  This creates an invalid state when:
  - User sets data_parallel_backend="ray" (to use Ray for DP)
  - AND async_scheduling=True (which requires distributed_executor_backend in ["mp", "uni"])

  ---
  📅 Timeline: How This Bug Survived Until Now

  | Date         | Event                                                        | Impact
                              |
  |--------------|--------------------------------------------------------------|--------------------------------------------------------------------------
  ----------------------------|
  | June 2, 2025 | https://github.com/vllm-project/vllm/pull/18779 merged       | 🐛 Bug introduced: data_parallel_backend="ray" incorrectly forces
  distributed_executor_backend="ray" |
  | Aug 29, 2025 | https://github.com/vllm-project/vllm/pull/23770 merged       | ⚠️ Test extended to cover async_scheduling=True, but bug remained hidden
                              |
  | Jun-Oct 2025 | Workaround existed in arg_utils.py                           | ✅ Early distributed_executor_backend="mp" assignment bypassed the bug
                               |
  | Oct 18, 2025 | https://github.com/vllm-project/vllm/commit/6367bde73 merged | 💥 Workaround removed → bug now exposed in CI
                              |

  ---
  ✅ The Fix

  Remove the incorrect coupling between data_parallel_backend and distributed_executor_backend:

  diff --git a/vllm/config/parallel.py b/vllm/config/parallel.py
  @@ -530,12 +530,6 @@ class ParallelConfig:
                           "ray`."
                       )
                   backend = "ray"
  -            elif self.data_parallel_backend == "ray":
  -                logger.info(
  -                    "Using ray distributed inference because "
  -                    "data_parallel_backend is ray"
  -                )
  -                backend = "ray"
               elif ray_found:
                   if self.placement_group:
                       backend = "ray"

  What this enables:
  - ✅ Ray for DP + mp for TP/PP (the failing test case)
  - ✅ mp for DP + Ray for TP/PP
  - ✅ Any other valid combination

  After this fix, data_parallel_backend="ray" will correctly use distributed_executor_backend="mp" (the default for single-node), allowing the test to
  pass.

  ---
  🙏 Request for Maintainers

  This is blocking the unrelated PR #26953 (GPTQ Marlin MoE mixed precision fixes).

  Option 1 (Preferred): Merge the 6-line fix to vllm/config/parallel.py shown above

  Option 2 (Temporary): Skip the failing test combination until this is resolved:
  if async_scheduling and data_parallel_backend == "ray":
      pytest.skip("Known issue: data_parallel_backend incorrectly forces distributed_executor_backend")

  cc @yewentao256 
  ```
  
  p.s. is using AI to summarize this bug helpful or not for situations like this? I felt it was better at explaining the root cause than me.

chriswritescode-dev commented:
looks ready to merge 👍 . Is there something else that is needed ?

mgoin commented on CMakeLists.txt:
```diff
@@ -668,9 +668,9 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
+    cuda_archs_loose_intersection(SCALED_MM_ARCHS "10.0a;12.0a" "${CUDA_ARCHS}")
   endif()
   if(${CMAKE_CUDA_COMPILER_VERSION} VERSION_GREATER_EQUAL 12.8 AND SCALED_MM_ARCHS)
     set(SRCS "csrc/quantization/w8a8/cutlass/moe/grouped_mm_c3x_sm100.cu")
```
This is not for marlin, it is for grouped_mm_c3x_sm100. Is this intentional somehow?

Commit:
Revert SM 12.0 support from SCALED_MM_ARCHS

Remove Blackwell architecture support from CMakeLists.txt as it is unrelated to GPTQ Marlin MoE functionality. Addresses review feedback from PR #26953.

Signed-off-by: CS <chris@chriswritescode.dev>

Commit:
Move workspace allocation from forward pass to init

Signed-off-by: CS <chris@chriswritescode.dev>

chriswritescode-dev committed:
Refactor Marlin MoE workspace allocation

- Centralize workspace size calculation in get_marlin_moe_workspace_size()
- Restore lazy allocation fallback for MXFP4 compatibility
- Optimize GPTQ path with pre-allocated workspace
- Add explicit workspace allocation to tests

Signed-off-by: Chris Scott <chris@getfit.us>

chriswritescode-dev committed:
Merge upstream/main into gptq-moe-fix

Resolve conflict in fused_marlin_moe.py by keeping both the new
get_marlin_moe_workspace_size() function and the upstream
default_activation_func() function.

Signed-off-by: Chris Scott <chris@getfit.us>

Commit:
Fix precommit issues: reorganize imports and shorten docstring

Signed-off-by: Chris Scott <chris@chriswritescode.dev>

chriswritescode-dev committed:
Merge branch 'main' into gptq-moe-fix

chriswritescode-dev committed:
Merge branch 'main' into gptq-moe-fix

Commit:
Centralize override_config call in get_moe_quant_method to eliminate redundant processing in GPTQMarlinMoEMethod constructor

Signed-off-by: Chris Scott <chris@chriswritescode.dev>

chriswritescode-dev committed:
Merge branch 'main' into gptq-moe-fix

chriswritescode-dev committed:
Merge branch 'main' into gptq-moe-fix

chriswritescode-dev committed:
Merge branch 'vllm-project:main' into gptq-moe-fix

chriswritescode-dev committed:
Merge branch 'main' into gptq-moe-fix

Commit:
Merge latest main into gptq-moe-fix

Resolve conflict in gptq_marlin.py by keeping enhanced metadata
handling from our PR and adding use_marlin flag from upstream.

Signed-off-by: Chris Scott <chris@chriswritescode.dev>

chriswritescode-dev committed:
Merge branch 'vllm-project:main' into gptq-moe-fix

mgoin commented:
Test failure seems related
```
[2025-11-16T06:54:59Z] FAILED quantization/test_gptq_marlin_moe_mixed_precision.py::test_gptq_marlin_moe_method_uses_dynamic_bits_override - assert 4 == 8
```

Commit:
Update GPTQ Marlin MoE mixed precision tests to use get_moe_quant_method

The tests were directly instantiating GPTQMarlinMoEMethod and manually
setting _layer_prefix, which bypassed the proper initialization flow
through get_moe_quant_method where override_config is called.

This update makes the tests use the intended API (get_moe_quant_method)
which properly applies dynamic quantization overrides, matching the
design where override_config is centralized in get_moe_quant_method.

Signed-off-by: CS <chris@chriswritescode.dev>

Commit:
Fix pre-commit issues: trailing whitespace and line endings

Signed-off-by: Chris Scott <chris@chriswritescode.dev>

Commit:
Merge upstream/main into gptq-moe-fix

Resolves conflicts:
- csrc/moe/marlin_moe_wna16/ops.cu: Keep b_thread_vecs fix with upstream's stages/is_a_8bit
- vllm/model_executor/layers/quantization/gptq_marlin.py: Keep _layer_prefix and _initialize_quant_metadata with upstream's input_dtype

Fixes:
- tests/kernels/moe/test_moe.py: quant_type -> b_type
- benchmarks/kernels/benchmark_moe.py: Use vllm.triton_utils import

Signed-off-by: Chris Scott <chris@chriswritescode.dev>
